src/main/java/com/epam/reportportal/cucumber/AbstractReporter.java

```diff
@@ -100,69 +128,56 @@ protected void beforeLaunch() {
     protected void afterLaunch() {
         FinishExecutionRQ finishLaunchRq = new FinishExecutionRQ();
         finishLaunchRq.setEndTime(Calendar.getInstance().getTime());
-        RP.get().finish(finishLaunchRq);
-    }
-
-    /**
-     * Manipulations before the feature starts
-     */
-    protected void beforeFeature() {
-        startFeature();
-    }
-
-    /**
-     * Finish Cucumber feature
-     */
-    protected void afterFeature() {
-        Utils.finishTestItem(RP.get(), currentFeatureContext.getFeatureId());
-        currentFeatureContext = null;
+        launch.get().finish(finishLaunchRq);
     }
 
     /**
      * Start Cucumber scenario
      */
-    protected void beforeScenario() {
-        Maybe<String> id = Utils.startNonLeafNode(RP.get(),
+    private void beforeScenario(RunningContext.FeatureContext currentFeatureContext,
+                                RunningContext.ScenarioContext currentScenarioContext, String scenarioName) {
+        Maybe<String> id = Utils.startNonLeafNode(
+                launch.get(),
                 currentFeatureContext.getFeatureId(),
-                Utils.buildNodeName(currentScenarioContext.getKeyword(), AbstractReporter.COLON_INFIX, currentScenarioContext.getName(), currentScenarioContext.getOutlineIteration()),
+                scenarioName,
                 currentFeatureContext.getUri() + ":" + currentScenarioContext.getLine(),
                 currentScenarioContext.getTags(),
-                getScenarioTestItemType()
-        );
+                getScenarioTestItemType());
         currentScenarioContext.setId(id);
     }
 
     /**
      * Finish Cucumber scenario
+     * Put scenario end time in a map to check last scenario end time per feature
      */
-    protected void afterScenario(TestCaseFinished event) {
-        Utils.finishTestItem(RP.get(), currentScenarioContext.getId(), event.result.getStatus().toString());
-        currentScenarioContext = null;
+    private void afterScenario(TestCaseFinished event) {
+        RunningContext.ScenarioContext currentScenarioContext = getCurrentScenarioContext();
+        for (Map.Entry<Pair<String, String>, RunningContext.ScenarioContext> scenarioContext : currentScenarioContextMap.entrySet()) {
+            if (scenarioContext.getValue().getLine() == currentScenarioContext.getLine()) {
+                currentScenarioContextMap.remove(scenarioContext.getKey());
```

Comment:
since you basically implement foreach here, i would suggest using iterator (that should have .remove method)